### ISSUE TITLE
[RFC] TorchVision with Batteries included - Phase 1

### ISSUE THREAD

**Issue by datumbox** [pytorch/vision#3911]
## 🚀 Feature

_Note: To track the progress of the project check out [this board](https://github.com/pytorch/vision/projects/1)._

Add popular primitives (Losses, Schedulers, Data Augmentations, Operators etc) which are often used to reproduce SOTA references and new popular highly accurate models with pre-trained weights to TorchVision.

## Motivation

Though TorchVision currently includes many common building blocks necessary for training CV models, it currently lacks popular primitives which are often used to reproduce SOTA. Some of these primitives are part of our reference scripts (Data utils, transforms etc) because previously did not want to commit to a specific API. Others are part of libraries from the broader ecosystem. Additionally, it does not provide some of the newer, popular architectures which currently achieve good results in a variety of vision tasks.

Adding support of such primitives and models to TorchVision will give a “batteries included” experience to its users. Researchers will be able to do SOTA research and reproduce papers by using common building blocks rather than rewriting their own while industry users will be able to adapt easier the models in their domains using SOTA techniques.

## Pitch

The addition of primitives should be done in several phases, iterating between trying to reproduce SOTA recipes, identifying accuracy gaps and implementing the necessary methods to close them. The progress of this project is tracked on [this board](https://github.com/pytorch/vision/projects/1).

During phase 1, add to TorchVision the following primitives and models:

- Losses #2980:
  - [x] [LabelSmoothing Loss](https://github.com/pytorch/pytorch/issues/7455) https://github.com/pytorch/pytorch/pull/63122 
  - [x] [SoftTarget CrossEntropy](https://github.com/pytorch/pytorch/issues/11959) https://github.com/pytorch/pytorch/pull/61044
- Schedulers:
  - [x] [ChainedScheduler](https://github.com/pytorch/pytorch/pull/26423#discussion_r329976246) https://github.com/pytorch/pytorch/pull/63491 https://github.com/pytorch/pytorch/pull/63457 https://github.com/pytorch/pytorch/pull/65034
  - [x] ConstantLR and LinearLR for [warmup](https://github.com/pytorch/pytorch/pull/60836) https://github.com/pytorch/pytorch/pull/64395
  - [x] [SequentialLR](https://github.com/pytorch/vision/issues/4281) https://github.com/pytorch/pytorch/pull/64037 https://github.com/pytorch/pytorch/pull/65035
- Models #2707:
  - [x] [EfficientNet (B0 to B7)](https://github.com/pytorch/vision/issues/980) #4293
  - [x] [RegNet](https://github.com/pytorch/vision/issues/2655) #4403
  - [x] [ViT](https://github.com/pytorch/vision/issues/4593) #4594
- Data Augmentations #3817: 
  - [x] [RandAugment](https://arxiv.org/abs/1909.13719v2) #4348
  - [x] [TrivialAugment](https://arxiv.org/abs/2103.10158) #4221 
  - [x] [MixUp](https://arxiv.org/abs/1710.09412) #4379
  - [x] [CutMix](https://arxiv.org/abs/1905.04899) #4379
- Operators:
  - [x] [Stochastic Depth](https://github.com/pytorch/vision/issues/4298) #4301
  - [x] [masks_to_boxes operator](https://github.com/pytorch/vision/issues/3960) #4290
- Training Recipes:
  - [x] [EMA model support](https://github.com/pytorch/vision/issues/4346) #4381 #4406 #4408
  - [x] [Updated reference scripts](https://github.com/pytorch/vision/issues/4281) #4335 #4411 #4444 #4493

Other potential primitives to be considered during phase 2:
* [Barron loss](https://arxiv.org/pdf/1701.03077.pdf) see [classy_vision](https://github.com/facebookresearch/ClassyVision/blob/main/classy_vision/losses/barron_loss.py).
* [Augmix + JSD loss](https://arxiv.org/abs/1912.02781)
* [FastAutoAugment](https://arxiv.org/abs/1905.00397)
* [Large Scale Jitter](https://github.com/facebookresearch/detectron2/blob/main/configs/new_baselines/mask_rcnn_R_50_FPN_100ep_LSJ.py#L44)
* [AutoDropout Layer](https://arxiv.org/abs/2101.01761)
* [DropBlock Layer](https://arxiv.org/abs/1810.12890)
* [DropConnect Layer](http://yann.lecun.com/exdb/publis/pdf/wan-icml-13.pdf)
* [ShakeDrop Layer](https://arxiv.org/abs/1802.02375)
* [Random Noise LR Scheduler](https://arxiv.org/abs/1810.01322)
* Truncated Normal in torch.distributions https://github.com/pytorch/pytorch/issues/2129 https://github.com/pytorch/pytorch/issues/32293

Note that any of the suggested primitives that are not vision-specific should be added on PyTorch, so that all Domain libraries can benefit from them.

cc @vfdev-5 @fmassa @oke-aditya @jbschlosser @iramazanli

**Comment by oke-aditya** [comment]
Adding one little bit. There have been such approaches to create batteries loaded libraries on top of torchvision.
We might as well take inspirations, motivations and add features that can be useful.

1. [Lightning bolts](https://github.com/PyTorchLightning/lightning-bolts): -

This uses PyTorch Lightning.
One can easily train and fine-tune models on distributed machines by simply inheriting classes. 
This saves lot of time of rewiring distributed scripts and facing issues.
E.g. FRCNN Training [script](https://github.com/PyTorchLightning/lightning-bolts/blob/master/pl_bolts/models/detection/faster_rcnn/faster_rcnn_module.py).
Also it is much easier to configure [backbones](https://github.com/PyTorchLightning/lightning-bolts/blob/master/pl_bolts/models/detection/faster_rcnn/backbones.py), it provides easier API.

2. [Kornia](https://kornia.readthedocs.io/en/latest/)

Well this has lot of features. I will highlight few important ones. Kornia is more like Open-cv with PyTorch.

`kornia.augmentations` 
This includes augmentations like in torchvision. But it also supports 3D transforms.
We can compare which could be useful additions to torchvision.

`kornia.color`
This provides common conversions for color scalse (rgb, hsv, bgr, grayscale, yuv, etc). This might be useful since `torchvision.io` can handle many image types.

`kornia.feature`. 
This is very similar to `torchvision.ops` providing 2D, 3D NMS. Also it has few classical computer vision algorithms such as SIFT, harris corner, Shit-tomasi etc. 
An interesting thing is `Matching` algorithms, which are used in say image similarity search, siamese networks, etc.

`kornia.losses`.
A loss functions module. Providing both functional as well as class based losses. This can be useful.

<details><summary>Some not very useful traditional CV features</summary>

`kornia.geometry`.
This reminds me of camera calibration! It helps to calculate epi-polar co-ordinates, fundamental matrix, etc.
Mostly used in traditional CV algorithms for structure to motion. (E.g. optical flow? Image Pyramids?) 

`kornia.morphology`.
Traditional CV algorithms for morphological operations. 
Here is kornia's future [roadmap](https://github.com/kornia/kornia/issues/930)

</details>

A higher level training API torchvision would be nice.
Most people want to just fine-tune their models, create baselines using torchvision. (We can discuss this in separate issue too)

**Comment by bmanga** [comment]
+1 from me.
Maybe there should be an experimental namespace where we can refine the API over time before promoting features to stable.

**Comment by oke-aditya** [comment]
Any thoughts about adding different types of IoU metrics such as DIoU (Distance IoU) and CIoU (Complete IoU)
Refer [paper](https://arxiv.org/abs/1911.08287) 

The above paper mentions some benefits in training with FRCNN, SSD and YOLOv4.
These are now used by [YOLOv5](https://github.com/ultralytics/yolov5/blob/master/utils/metrics.py)

This was earlier asked in #3026 #2545

Probably these two operations are more mature now to be included?

**Comment by datumbox** [comment]
@oke-aditya I just realized I haven't responded. Apologies for that.

I agree we should explore the ideas you added. I know you have separate issues for all of them, so let's track them there. Happy to move them to Batteries Included once the first set of primitives is added. Some additional discussion will be necessary to prioritize but we'll do this when it's time to review additions.

**Comment by oke-aditya** [comment]
Hey, now worries @datumbox  :smile: You are doing an awesome work, and batteries included is a great initiative :+1: 

Let me know if there are any features that I could work on. Maybe ones that are not critical or timeline specific so that it won't hinder overall development.

**Comment by datumbox** [comment]
@oke-aditya I'm currently collecting feedback from various research teams at FB to see which other operators are worth including. I will definitely let you know once things become clear; I just want to make sure you won't work on something that we later feel shouldn't be added. Any case if you want to have an early look, check out the [DropBlock](https://arxiv.org/pdf/1810.12890.pdf) paper.

**Comment by vaibhava0** [comment]
I recommend adding `Large Scale Jitter` data augmentation as well. It is quite simple and powerful. It has shown promise across many use cases. Here is an implementation in D2:

https://github.com/facebookresearch/detectron2/blob/main/configs/new_baselines/mask_rcnn_R_50_FPN_100ep_LSJ.py#L44

Some benchmark results:
https://github.com/facebookresearch/detectron2/blob/main/MODEL_ZOO.md#new-baselines-using-large-scale-jitter-and-longer-training-schedule

**Comment by 0x00b1** [comment]
Oof! I feel bad! This is shaping up to be a great release.

**Comment by datumbox** [comment]
@vaibhava0 Thanks for the proposal. You are right, this is quite important. We have it on the https://github.com/pytorch/vision/issues/3817 but I'll add it here more prominently (I also added the code and the benchmarks links you provided). 

@0x00b1 Your contribution can still make it in. tap tap tap ⌨️  🚀

**Comment by oke-aditya** [comment]
> Any thoughts about adding different types of IoU metrics such as DIoU (Distance IoU) and CIoU (Complete IoU)
> Refer [paper](https://arxiv.org/abs/1911.08287)

CIoU and DIoU have been added to Detectron2

https://github.com/facebookresearch/detectron2/blob/dfe8d368c8b7cc2be42c5c3faf9bdcc3c08257b1/detectron2/layers/losses.py#L66

**Comment by datumbox** [comment]
Batteries Included - Phase 1 is now concluded! I believe we have successfully refreshed TorchVision library to support the Classification use-case and we have managed to refresh all of the popular pre-trained models of the library. 

I'm going to close this ticket and start scoping Phase 2, which will focus on the Detection and Segmentation use-cases. Massive thanks to all of the people who contributed to this project either by implementing primitives, adding models or training new weights. I'll follow up with a new RFC for phase 2 and start outlining next steps so that we can get the feedback from the community.

**Comment by xiaoyuan0203** [comment]
Will Mixup and CutMix be added to `torchvision.transform` and when?  #4379

**Comment by datumbox** [comment]
@xiaoyuan0203 that's the plan. We have untested implementations on [prototype](https://github.com/pytorch/vision/blob/main/torchvision/prototype/transforms/_augment.py#L104-L177). We are working to finalize the API, document them and start testing them. I don't recommend them yet for production use-cases, but we will make sure to post an update when our confidence in them is increased.

**Comment by laurentperrinet** [comment]
Note: The board linked to https://github.com/pytorch/vision/projects/1 is not accessible.
